### DISCUSSION TITLE
(match) is not strict

### DISCUSSION THREAD

**Discussion by iacore** [janet-lang/janet#1273] [General]
`(match [1] [] 1 0)` is `1`.

What currently works:

```
(match [1] ([& xs] (empty? xs)) 1 0)
```

**Comment by CosmicToast** [comment]
Yes, match is strict on types and then on values.
If you don't specify a value, match will simply match on the type.
However, match statements are *ordered*, so I recommend doing something along these lines:
```janet
(match [:some]
  [a b & rest] :two-or-more
  [a] :one # depending on your use case just [head & tail] is sufficient, e.g for recursive algorithms
  [] :empty)
```

**Reply by iacore** [reply]
this is not pattern matching that I expected :laughing:

**Reply by sogaiu** [reply]
You are not alone :)

FWIW, there is some coverage of some unexpectedness of `match` in the [Control FLow chapter](https://janet.guide/control-flow/) of the "Janet for Mortals" book.

> Janet’s implementation of pattern matching happens to have a couple rough edges that you’ll need to be aware of when you’re using match.

If interested, search for "rough edges" at the page linked to.

**Reply by ianthehenry** [reply]
I wrote https://github.com/ianthehenry/pat because of this and the lack of alternative cases in the default `match`. It's *close* to a drop-in replacement, but with exact-by-default indexed patterns.
